### PR TITLE
[IN-2675] Flutter to FVM transition

### PR DESCRIPTION
**Pull request by nagypeterjob**
### New Pull Request Checklist

*Please mark the points which you did / accept.*

- [ ] The tool I added is stable, and __does not require frequent updates__. _Preinstalled tools on the LTS stacks
  are not updated, so if the tool requires frequent updates you should handle the installation on-demand (see: [Install Any Additional Tool - DevCenter](https://bitrise-io.github.io/devcenter/tips-and-tricks/install-additional-tools/) for more information)_
- [ ] I updated the corresponding tests if there were any.
- [ ] I added a version report line to [`system_report.sh`](/system_report.sh) for the new tool(s) I added
- [ ] I have added an entry to the CHANGELOG.md with a revisionID (the current date and the number of the change e.g.: 2019_10_11_1) and a brief description of the actual change
- [ ] I have updated the revisionID in [bitrise_profile](roles/profiles/files/bitrise_profile) **BITRISE_OSX_STACK_REV_ID** with the one I just created in the CHANGELOG.md

### CODE REVIEW

File: roles/flutter/tasks/main.yml [hunk 1]
```diff
@@ -37,18 +9,87 @@
...
       mode: 0755
 
-  - name: Generate script to profile.d
+  - name: Copy script to profile.d
     become: true
-    template:
-      src: patch.j2
+    copy:
+      src: flutter.sh
       dest: "~{{ param_user }}/profile.d/flutter.sh"
```
Comment (nagypeterjob) [inline roles/flutter/tasks/main.yml:16]: Can we lose the tilde & go with OS specific user location?

File: roles/flutter/tasks/main.yml [hunk 2]
```diff
@@ -37,18 +9,87 @@
...
+- name: Is flutter available?
+  shell: bash -l -c "flutter --version"
+  ignore_errors: true
+  register: is_flutter_available
+  changed_when: false
+
+- name: Remove old flutter
+  become: true
+  file: 
+    path: "/usr/local/flutter"
```
Comment (nagypeterjob) [inline roles/flutter/tasks/main.yml:33]: I am pretty sure that on MacOS, the correct location would be /usr/local/bin/flutter

File: roles/flutter/tasks/main.yml [hunk 3]
```diff
@@ -37,18 +9,87 @@
...
+- name: Install Flutter versions
+  include_tasks:
+    file: flutter.yml
+  with_items:
+    - "{{ flutter_versions }}"
+  loop_control:
+    loop_var: flutter_version
+
+- name: Set Flutter 1 as default
+  shell: bash -l -c "/home/linuxbrew/.linuxbrew/bin/fvm global {{ flutter_versions[0] }}" # complete version is required
```
Comment (nagypeterjob) [inline roles/flutter/tasks/main.yml:86]: can we put /home/linuxbrew/.linuxbrew/bin/fvm to a variable?

**Comment by nagypeterjob** [comment]
What I don't understand: Is this a linux only role? I don't see any indication of that

File: roles/flutter/tasks/main.yml (hunk 2 above)
Comment (peterheja-bitrise) [reply roles/flutter/tasks/main.yml:33]: Yup, plus `homebrew uninstall` is the clean version
But shouldn't break how it breaks now

File: roles/flutter/tasks/main.yml (hunk 1 above)
Comment (martonszasz-bitrise) [reply roles/flutter/tasks/main.yml:16]: As discussed with @fehersanyi-bitrise, paths are going to be the subject of final cleanup after we got this thing working, but sure.

File: roles/flutter/tasks/main.yml (hunk 3 above)
Comment (peterheja-bitrise) [reply roles/flutter/tasks/main.yml:86]: https://github.com/bitrise-io/osx-box-bootstrap/pull/484/files#diff-45ba7edcfad8e75d839dd9602f79eef79dfb4f25c40244609b17ef822d2df8c4R4 PATH is setup correctly

File: roles/flutter/tasks/main.yml (hunk 2 above)
Comment (martonszasz-bitrise) [reply roles/flutter/tasks/main.yml:33]: Same as before - paths are to be fixed / integrated into both MacOS and Linux envs after I got a clear run.

Comment (peterheja-bitrise) [reply roles/flutter/tasks/main.yml:33]: Sorry, I was wrong here.

File: roles/flutter/tasks/main.yml (hunk 3 above)
Comment (peterheja-bitrise) [reply roles/flutter/tasks/main.yml:86]: Although, yeah, we should set in flutter profile to point to the right flutter, e.g macOS export this to PATH: `/Users/vagrant/fvm/default/bin` which is basically `$HOME/fvm/default/bin`

Comment (peterheja-bitrise) [reply roles/flutter/tasks/main.yml:86]: I might even go as far as that Linuxbrew might not be properly supported with `fvm` - that's not so surreal either.

Comment (peterheja-bitrise) [reply roles/flutter/tasks/main.yml:86]: Btw, on a box:
```vagrant@prd-atl-std-y-xcode-12-4-210520131845-s5fmNZPvpTaP4x2SNFGtUn ~ % export PATH="/Users/vagrant/fvm/default/bin:${PATH}"
vagrant@prd-atl-std-y-xcode-12-4-210520131845-s5fmNZPvpTaP4x2SNFGtUn ~ % flutter --version
Flutter 1.26.0-17.8.pre • channel unknown • unknown source
Framework • revision 044f2cf560 (3 months ago) • 2021-02-24 13:02:05 -0800
Engine • revision 042c82b02c
Tools • Dart 2.12.0 (build 2.12.0-259.16.beta)
vagrant@prd-atl-std-y-xcode-12-4-210520131845-s5fmNZPvpTaP4x2SNFGtUn ~ % fvm list | grep 2.0.5
2.0.5
vagrant@prd-atl-std-y-xcode-12-4-210520131845-s5fmNZPvpTaP4x2SNFGtUn ~ % fvm global 2.0.5
Flutter "2.0.5" has been set as global
vagrant@prd-atl-std-y-xcode-12-4-210520131845-s5fmNZPvpTaP4x2SNFGtUn ~ % flutter --version
Flutter 2.0.5 • channel unknown • unknown source
Framework • revision adc687823a (5 weeks ago) • 2021-04-16 09:40:20 -0700
Engine • revision b09f014e96
Tools • Dart 2.12.3
```

So it seems to work, if PATH is correct.

File: roles/flutter/vars/main.yml [hunk 4]
```diff
@@ -0,0 +1,4 @@
+---
+BREW_HOME: "{{ \"/home/linuxbrew/.linuxbrew/bin\" if (ansible_os_family == 'Debian') else \"/usr/local/bin\" if (ansible_os_family == 'Darwin') else \"\" }}"
+FVM_HOME: "{{ \"/home/linuxbrew/fvm/default/bin\" if (ansible_os_family == 'Debian') else \"~/fvm/default/bin\" if (ansible_os_family == 'Darwin') else \"\" }}"
```
Comment (nagypeterjob) [inline roles/flutter/vars/main.yml:3]: one last suggestion: can we loose the ~ here?

File: roles/flutter/files/flutter.sh [hunk 5]
```diff
@@ -0,0 +1,5 @@
+export PATH="$PATH":"$HOME/.pub-cache/bin"
+# export PATH="$PATH":"$HOME/fvm/default/bin"
```
Comment (fehersanyi-bitrise) [inline roles/flutter/files/flutter.sh:2]: why is this commented out?

File: roles/flutter/tasks/main.yml [hunk 6]
```diff
@@ -37,18 +9,84 @@
...
+    path: "/usr/local/flutter"
+    state: absent
+
+- name: Is dart available?
+  shell: bash -l -c "dart --version"
+  ignore_errors: true
+  register: is_dart_available
+  changed_when: false
+
+- name: Tap and install dart
```
Comment (fehersanyi-bitrise) [inline roles/flutter/tasks/main.yml:39]: if we brew install flutter then I don't think this will be necessary

**Review by fehersanyi-bitrise — CHANGES_REQUESTED** [review]

File: roles/flutter/vars/main.yml (hunk 4 above)
Comment (peterheja-bitrise) [reply roles/flutter/vars/main.yml:3]: You mean replace it with a `$HOME` or similar?

File: roles/flutter/tasks/main.yml (hunk 6 above)
Comment (peterheja-bitrise) [reply roles/flutter/tasks/main.yml:39]: fvm depends on dart

File: roles/flutter/vars/main.yml (hunk 4 above)
Comment (nagypeterjob) [reply roles/flutter/vars/main.yml:3]: exactly

File: roles/flutter/tasks/main.yml [hunk 7]
```diff
@@ -37,18 +9,61 @@
       owner: "{{ param_user }}"
       mode: 0755
 
-  - name: Generate script to profile.d
+  - name: Copy script to profile.d
     become: true
     template:
-      src: patch.j2
+      src: flutter.sh.j2
       dest: "~{{ param_user }}/profile.d/flutter.sh"
```
Comment (nagypeterjob) [inline roles/flutter/tasks/main.yml:16]: I would really like this ~ to be gone, but other than that looks ok

**Comment by nagypeterjob** [comment]
Fine by me, I cannot approve my own PR, so @martonszasz-bitrise you can approve it for me :D

**Comment by fehersanyi-bitrise** [comment]
The test is still failing tho 🤔

**Comment by martonszasz-bitrise** [comment]
Yes, still trying to find out why the `flutter global` step fails in the test env, when it sets the global flutter in local just fine.

**Pull request closed without merging**